flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -5581,10 +5628,210 @@ class ArrayExprLowering {
+  CC applyPathToArrayLoad(mlir::Location loc, fir::ArrayLoadOp load) {
+    auto revPath = reversePath; // Force a copy to be made.
+    if (semant == ConstituentSemantics::ProjectedCopyInCopyOut) {
+      destination = load;
+      auto innerArg = explicitSpace->findArgumentOfLoad(load);
+      return [=](IterSpace iters) mutable {
+        auto [path, eleTy] = lowerPath(loc, revPath, load.getType(), iters);
+        auto toTy = adjustedArrayElementType(eleTy);
+        auto castedElement =
+            builder.createConvert(loc, toTy, iters.getElement());
+        auto update = builder.create<fir::ArrayUpdateOp>(
+            loc, innerArg.getType(), innerArg, castedElement, path,
+            load.typeparams());
+        // Flag the offsets as "Fortran" as they are not zero-origin.
+        update->setAttr(fir::factory::attrFortranArrayOffsets(),
+                        builder.getUnitAttr());
+        return arrayLoadExtValue(builder, loc, load, {}, update);
+      };
+    }
+    return [=](IterSpace iters) mutable {
+      auto [path, eleTy] = lowerPath(loc, revPath, load.getType(), iters);
+      auto resTy = adjustedArrayElementType(eleTy);
+      // TODO: When loading a subobject array, convert the fetch (value) to a
+      // fetch reference. This reference can then be used in a secondary
+      // array_load. Consider handling this another way.
+      if (semant == ConstituentSemantics::RefOpaque &&
+          !fir::isa_ref_type(resTy))
+        resTy = builder.getRefType(resTy);
+      auto fetch = builder.create<fir::ArrayFetchOp>(loc, resTy, load, path,
+                                                     load.typeparams());
+      // Flag the offsets as "Fortran" as they are not zero-origin.
+      fetch->setAttr(fir::factory::attrFortranArrayOffsets(),
+                     builder.getUnitAttr());
+      return arrayLoadExtValue(builder, loc, load, path, fetch);
+    };
+  }
+  CC applyPathToArrayLoad(fir::ArrayLoadOp load) {
+    auto loc = getLoc();
+    auto lambda = applyPathToArrayLoad(loc, load);
+    reversePath.clear();
+    return lambda;
+  }
+
+  CC genexp(const Fortran::semantics::Symbol &x) {
+    if (auto load = explicitSpace->findBinding(&x))
+      return applyPathToArrayLoad(load);
+    if (pathIsEmpty())
+      return [=](IterSpace) { return asScalar(x); };
+    return [=](IterSpace) { return ExtValue{}; };
   }
 
+  CC genexp(const Fortran::evaluate::Component &x) {
+    if (auto load = explicitSpace->findBinding(&x))
+      return applyPathToArrayLoad(load);
+    auto top = pathIsEmpty();
+    reversePath.push_back(&x);
+    auto result = genexp(x.base());
+    if (pathIsEmpty())
+      return result;
+    if (top)
+      return [=](IterSpace) { return asScalar(x); };
+    return [=](IterSpace) { return ExtValue{}; };
+  }
+
+  CC genexp(const Fortran::evaluate::ArrayRef &x) {
+    if (auto load = explicitSpace->findBinding(&x)) {
+      if (x.subscript().empty()) {
```

Comment:
It seems like I stumbled upon a case. Is it asserted somewhere in evaluate to be impossible?

Reply:
> It seems like I stumbled upon a case. Is it asserted somewhere in evaluate to be impossible?

No I do not think it is asserted. However any unparsed with an arrayRef with no subscripts would just be bad (it would look like `array)` given ArrayRef [AsFortran](https://github.com/llvm/llvm-project/blob/9245f35580ca0cce147ec9cebfa431fa5b7feac4/flang/lib/Evaluate/formatting.cpp#L649) function). So if there is any ArrayRef without susbcripts, I think it's a bug.

I asked @klausler if this is the case and if we could add a CHECK there.